---
layout: post
title:  "Virome: a primer"
author: at
categories: [ virome, tutorial]
image: assets/images/virome.jpg
hidden: false
---

"Virome refers to the assemblage of viruses that is often investigated and described by metagenomic sequencing of viral nucleic acids that are found associated with a particular ecosystem, organism or holobiont. The word is frequently used to describe environmental viral shotgun metagenomes." ([Wikipedia](https://en.wikipedia.org/wiki/Virome))

## Reads profiling

Nowadays there are enough viral genomes deposited in the public repositories
to allow us to profile the reads of a metagenomic sample against them. 
This approach has the advantages of being fast and - when used with separate datasets - 
enables a straightforward comparison of the results. 

The main disadvantage is that novel viruses will not be detected at all, and they might play an
important role in our community.

## *De novo* virus mining

A different approach is to assemble our reads and then to try to identify which contigs 
could be complete or partial viral genomes.
This approach can detect novel viruses, but can also be prone to false positives and requires
some refinement.

In addition to this, the lack of universal markers makes the taxonomy assignment more complicated
and we will see how to use vConTACT2 to investigate the relationship of the detected viral sequences
with the known genomes.



---

## The programme

* :one: [Gathering the reads]({{ site.baseurl }}{% link _posts/2022-02-10-virome-reads.md %}):
  downloading and subsampling reads from public repositories (optional)
* :two: [Gathering the tools]({{ site.baseurl }}{% link _posts/2022-02-11-virome-tools.md %}):
  we will use Miniconda to manage our dependencies
* :three: [Reads by reads profiling]({{ site.baseurl }}{% link _posts/2022-02-12-virome-phanta.md %}):
  using Phanta to quickly profile the bacterial and viral components of a microbial community
* :four:  [_De novo_ mining]({{ site.baseurl }}{% link _posts/2022-02-13-virome-denovo.md %}):
  assembly based approach, using VirSorter as an example miner
* :five:  [_De novo_ mining]({{ site.baseurl }}{% link _posts/2022-02-14-virome-taxonomy.md %}):
  *ab initio* taxonomy profiling using vConTACT2
* :six:  [MetaPhage overview]({{ site.baseurl }}{% link _posts/2022-02-15-virome-metaphage.md %}):
  what is MetaPhage, a reads to report pipeline for viral metagenomics

:arrow_left: [Back to the main page]({{ site.baseurl }}{% link _posts/2022-02-01-Virome.md %})